Should not rename structs in standard library

<!--
Troubleshooting guide: https://rust-analyzer.github.io/manual.html#troubleshooting
Forum for questions: https://users.rust-lang.org/c/ide/14

Before submitting, please make sure that you're not running into one of these known issues:

 1. on-the-fly diagnostics are mostly unimplemented (`cargo check` diagnostics will be shown when saving a file): #3107

Otherwise please try to provide information which will help us to fix the issue faster. Minimal reproducible examples with few dependencies are especially lovely <3.
-->

I accidentally renamed `Option::None` to `NoOp`. Although I closed my editor (Neovim) immediately, the release build was failing.

<details>

```sh
$ cargo b --release
   Compiling aho-corasick v1.0.1
   Compiling errno v0.3.1
   Compiling io-lifetimes v1.0.10
   Compiling syn v1.0.109
   Compiling log v0.4.17
   Compiling once_cell v1.17.1
   Compiling termcolor v1.2.0
error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.17.1/src/lib.rs:444:47
    |
444 |             OnceCell { inner: UnsafeCell::new(NoOp) }
    |                                               ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.17.1/src/lib.rs:936:17
    |
936 |                 NoOp
    |                 ^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
error: could not compile `once_cell` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:589:18
    |
589 |             _ => NoOp,
    |                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:768:18
    |
768 |             _ => NoOp,
    |                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:943:18
    |
943 |             _ => NoOp,
    |                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:927:24
    |
927 |             separator: NoOp,
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:958:18
    |
958 |             _ => NoOp,
    |                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:1662:23
     |
1662 |             fg_color: NoOp,
     |                       ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:1058:30
     |
1058 |                 module_path: NoOp,
     |                              ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:1663:23
     |
1663 |             bg_color: NoOp,
     |                       ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:1844:25
     |
1844 |         self.fg_color = NoOp;
     |                         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:1845:25
     |
1845 |         self.bg_color = NoOp;
     |                         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:1059:23
     |
1059 |                 file: NoOp,
     |                       ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:1060:23
     |
1060 |                 line: NoOp,
     |                       ^^^^ not found in this scope

error: could not compile `log` due to 7 previous errors
error: could not compile `termcolor` due to 5 previous errors
error[E0408]: variable `NoOp` is not bound in all patterns
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:251:20
    |
251 |             NoOp | Some(ForceAlgorithm::Teddy) => {
    |             ----   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ pattern doesn't bind `NoOp`
    |             |
    |             variable not in all patterns
    |
help: if you meant to match on a variant or a `const` item, consider making the path in the pattern qualified: `path::to::ModOrType::NoOp`
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:251:13
    |
251 |             NoOp | Some(ForceAlgorithm::Teddy) => {
    |             ^^^^

error[E0416]: identifier `NoOp` is bound more than once in the same pattern
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:207:20
    |
207 |             (NoOp, NoOp) if self.ascii_case_insensitive => NoOp,
    |                    ^^^^ used in a pattern more than once

error[E0416]: identifier `NoOp` is bound more than once in the same pattern
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:208:20
    |
208 |             (NoOp, NoOp) => {
    |                    ^^^^ used in a pattern more than once

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:816:33
    |
816 |         OverlappingState { mat: NoOp, id: NoOp, at: 0, next_match_index: NoOp }
    |                                 ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:816:43
    |
816 |         OverlappingState { mat: NoOp, id: NoOp, at: 0, next_match_index: NoOp }
    |                                           ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:816:74
    |
816 |         OverlappingState { mat: NoOp, id: NoOp, at: 0, next_match_index: NoOp }
    |                                                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:869:51
    |
869 |         Ok(FindIter { aut, input, last_match_end: NoOp })
    |                                                   ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1172:32
     |
1172 |                         return NoOp;
     |                                ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1209:9
     |
1209 |         NoOp
     |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1224:9
     |
1224 |         NoOp
     |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1235:9
     |
1235 |         NoOp
     |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1264:19
     |
1264 |         return Ok(NoOp);
     |                   ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1268:38
     |
1268 |         try_find_fwd_imp(aut, input, NoOp, Anchored::Yes, earliest)
     |                                      ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1277:42
     |
1277 |             try_find_fwd_imp(aut, input, NoOp, Anchored::No, true)
     |                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1279:42
     |
1279 |             try_find_fwd_imp(aut, input, NoOp, Anchored::No, false)
     |                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1294:19
     |
1294 |     let mut mat = NoOp;
     |                   ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1303:42
     |
1303 |             Candidate::None => return Ok(NoOp),
     |                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1428:17
     |
1428 |     state.mat = NoOp;
     |                 ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1438:50
     |
1438 |         try_find_overlapping_fwd_imp(aut, input, NoOp, state)
     |                                                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1485:42
     |
1485 |                 state.next_match_index = NoOp;
     |                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1486:29
     |
1486 |                 state.mat = NoOp;
     |                             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1586:46
     |
1586 |     let mut cur: Option<(u8, u8, StateID)> = NoOp;
     |                                              ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1606:9
     |
1606 |         NoOp
     |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/dfa.rs:770:26
    |
770 |     let mut prev_class = NoOp;
    |                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/contiguous.rs:823:28
    |
823 |                     return NoOp;
    |                            ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/contiguous.rs:836:21
    |
836 |                     NoOp
    |                     ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/contiguous.rs:841:28
    |
841 |                     return NoOp;
    |                            ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/noncontiguous.rs:581:28
    |
581 |                 prefilter: NoOp,
    |                            ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/noncontiguous.rs:1243:26
     |
1243 |         QueuedSet { set: NoOp }
     |                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:117:20
    |
117 |             force: NoOp,
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:118:30
    |
118 |             force_teddy_fat: NoOp,
    |                              ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:119:24
    |
119 |             force_avx: NoOp,
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:145:26
    |
145 |             self.force = NoOp;
    |                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:183:26
    |
183 |             self.force = NoOp;
    |                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:240:20
    |
240 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:644:20
    |
644 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/pattern.rs:213:20
    |
213 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/rabinkarp.rs:111:20
    |
111 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/rabinkarp.rs:124:24
    |
124 |                 return NoOp;
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/rabinkarp.rs:164:13
    |
164 |             NoOp
    |             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
  --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/teddy/mod.rs:24:13
   |
24 |             NoOp
   |             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
  --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/teddy/mod.rs:46:13
   |
46 |             NoOp
   |             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
  --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/alphabet.rs:94:73
   |
94 |         ByteClassElementRanges { elements: self.elements(class), range: NoOp }
   |                                                                         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/alphabet.rs:154:9
    |
154 |         NoOp
    |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
  --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:89:32
   |
89 |             Candidate::None => NoOp,
   |                                ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:165:20
    |
165 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:283:24
    |
283 |             self.one = NoOp;
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:421:13
    |
421 |             NoOp
    |             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:457:24
    |
457 |                 return NoOp;
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:467:29
    |
467 |                 0 => return NoOp,
    |                             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:696:24
    |
696 |                 return NoOp;
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:710:28
    |
710 |                     return NoOp;
    |                            ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:716:29
    |
716 |                 0 => return NoOp,
    |                             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/primitives.rs:374:20
    |
374 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/search.rs:1105:36
     |
1105 |             MatchKind::Standard => NoOp,
     |                                    ^^^^ not found in this scope

Some errors have detailed explanations: E0408, E0416, E0425.
For more information about an error, try `rustc --explain E0408`.
error: could not compile `aho-corasick` due to 55 previous errors
error: could not compile `aho-corasick` due to 55 previous errors
error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.109/src/lib.rs:967:23
    |
967 |     let mut shebang = NoOp;
    |                       ^^^^ not found in this scope

error: could not compile `syn` due to previous error
```

</details>

To solve this, I deleted my `~/.cargo/registry`.

However, I believe that Rust-Analyzer should have prevented me from renaming `Option::None` in the first place because it is a struct defined in the standard library.

**rust-analyzer version**: rust-analyzer 0.3.1402-standalone (646f97385 2023-02-12)

**rustc version**: rustc 1.69.0 (84c898d65 2023-04-16)

--- a/crates/ide-db/src/rename.rs
+++ b/crates/ide-db/src/rename.rs
@@ -22,7 +22,7 @@
 //! Our current behavior is ¯\_(ツ)_/¯.
 use std::fmt;
 
-use base_db::{AnchoredPathBuf, FileId, FileRange};
+use base_db::{AnchoredPathBuf, CrateOrigin, FileId, FileRange};
 use either::Either;
 use hir::{FieldSource, HasSource, InFile, ModuleSource, Semantics};
 use stdx::never;
@@ -77,6 +77,15 @@ impl Definition {
                 bail!("Cannot rename builtin type")
             }
             Definition::SelfType(_) => bail!("Cannot rename `Self`"),
+            Definition::Adt(hir::Adt::Struct(strukt)) => {
+                if !matches!(
+                    strukt.module(sema.db).krate().origin(sema.db),
+                    CrateOrigin::Local { .. }
+                ) {
+                    bail!("Cannot rename a non-local struct.")
+                }
+                rename_reference(sema, Definition::Adt(hir::Adt::Struct(strukt)), new_name)
+            }
             def => rename_reference(sema, def, new_name),
         }
     }

--- a/crates/ide-db/src/rename.rs
+++ b/crates/ide-db/src/rename.rs
@@ -93,13 +93,7 @@ impl Definition {
                 bail!("Cannot rename a builtin attr.")
             }
             Definition::SelfType(_) => bail!("Cannot rename `Self`"),
-            Definition::Macro(mac) => {
-                if mac.is_builtin_derive(sema.db) {
-                    bail!("Cannot rename builtin derive")
-                }
-
-                rename_reference(sema, Definition::Macro(mac), new_name)
-            }
+            Definition::Macro(mac) => rename_reference(sema, Definition::Macro(mac), new_name),
             def => rename_reference(sema, def, new_name),
         }
     }

--- a/crates/ide-db/src/rename.rs
+++ b/crates/ide-db/src/rename.rs
@@ -73,6 +73,8 @@ impl Definition {
     ) -> Result<SourceChange> {
         // self.krate() returns None if
         // self is a built-in attr, built-in type or tool module.
+        // it is not allowed for these defs to be renamed.
+        // cases where self.krate() is None is handled below.
         if let Some(krate) = self.krate(sema.db) {
             if !krate.origin(sema.db).is_local() {
                 bail!("Cannot rename a non-local definition.")

--- a/crates/ide-db/src/rename.rs
+++ b/crates/ide-db/src/rename.rs
@@ -71,20 +71,24 @@ impl Definition {
         sema: &Semantics<'_, RootDatabase>,
         new_name: &str,
     ) -> Result<SourceChange> {
+        if let Some(krate) = self.krate(sema.db) {
+            if !matches!(krate.origin(sema.db), CrateOrigin::Local { .. }) {
+                bail!("Cannot rename a non-local definition.")
+            }
+        }
+
         match *self {
             Definition::Module(module) => rename_mod(sema, module, new_name),
             Definition::BuiltinType(_) => {
                 bail!("Cannot rename builtin type")
             }
             Definition::SelfType(_) => bail!("Cannot rename `Self`"),
-            Definition::Adt(hir::Adt::Struct(strukt)) => {
-                if !matches!(
-                    strukt.module(sema.db).krate().origin(sema.db),
-                    CrateOrigin::Local { .. }
-                ) {
-                    bail!("Cannot rename a non-local struct.")
+            Definition::Macro(mac) => {
+                if mac.is_builtin_derive(sema.db) {
+                    bail!("Cannot rename builtin macro")
                 }
-                rename_reference(sema, Definition::Adt(hir::Adt::Struct(strukt)), new_name)
+
+                rename_reference(sema, Definition::Macro(mac), new_name)
             }
             def => rename_reference(sema, def, new_name),
         }

--- a/crates/ide-db/src/rename.rs
+++ b/crates/ide-db/src/rename.rs
@@ -22,7 +22,7 @@
 //! Our current behavior is ¯\_(ツ)_/¯.
 use std::fmt;
 
-use base_db::{AnchoredPathBuf, CrateOrigin, FileId, FileRange};
+use base_db::{AnchoredPathBuf, FileId, FileRange};
 use either::Either;
 use hir::{FieldSource, HasSource, InFile, ModuleSource, Semantics};
 use stdx::never;
@@ -71,21 +71,29 @@ impl Definition {
         sema: &Semantics<'_, RootDatabase>,
         new_name: &str,
     ) -> Result<SourceChange> {
+        // self.krate() returns None if
+        // self is a built-in attr, built-in type or tool module.
         if let Some(krate) = self.krate(sema.db) {
-            if !matches!(krate.origin(sema.db), CrateOrigin::Local { .. }) {
+            if !krate.origin(sema.db).is_local() {
                 bail!("Cannot rename a non-local definition.")
             }
         }
 
         match *self {
             Definition::Module(module) => rename_mod(sema, module, new_name),
+            Definition::ToolModule(_) => {
+                bail!("Cannot rename a tool module")
+            }
             Definition::BuiltinType(_) => {
                 bail!("Cannot rename builtin type")
+            }
+            Definition::BuiltinAttr(_) => {
+                bail!("Cannot rename a builtin attr.")
             }
             Definition::SelfType(_) => bail!("Cannot rename `Self`"),
             Definition::Macro(mac) => {
                 if mac.is_builtin_derive(sema.db) {
-                    bail!("Cannot rename builtin macro")
+                    bail!("Cannot rename builtin derive")
                 }
 
                 rename_reference(sema, Definition::Macro(mac), new_name)

--- a/crates/ide/src/rename.rs
+++ b/crates/ide/src/rename.rs
@@ -2639,7 +2639,7 @@ use qux as frob;
 =======
 
     #[test]
-    fn disallow_renaming_for_non_local_struct() {
+    fn disallow_renaming_for_non_local_definition() {
         check(
             "Baz",
             r#"
@@ -2648,7 +2648,7 @@ pub struct S$0;
 //- /main.rs crate:main deps:lib new_source_root:local
 use lib::S;
 "#,
-            "error: Cannot rename a non-local struct.",
+            "error: Cannot rename a non-local definition.",
         );
     }
 >>>>>>> 948d9f274 (Disallow renaming of non-local structs)

--- a/crates/ide/src/rename.rs
+++ b/crates/ide/src/rename.rs
@@ -2529,7 +2529,6 @@ fn main() {
 ",
         )
     }
-<<<<<<< HEAD
 
     #[test]
     fn extern_crate() {
@@ -2635,8 +2634,6 @@ use qux as frob;
         // ",
         //         );
     }
-||||||| parent of 948d9f274 (Disallow renaming of non-local structs)
-=======
 
     #[test]
     fn disallow_renaming_for_non_local_definition() {
@@ -2644,12 +2641,26 @@ use qux as frob;
             "Baz",
             r#"
 //- /lib.rs crate:lib new_source_root:library
-pub struct S$0;
+pub struct S;
 //- /main.rs crate:main deps:lib new_source_root:local
-use lib::S;
+use lib::S$0;
 "#,
             "error: Cannot rename a non-local definition.",
         );
     }
->>>>>>> 948d9f274 (Disallow renaming of non-local structs)
+
+    #[test]
+    fn disallow_renaming_for_builtin_macros() {
+        check(
+            "Baz",
+            r#"
+//- minicore: derive, hash
+//- /main.rs crate:main
+use core::hash::Hash;
+#[derive(H$0ash)]
+struct A;
+            "#,
+            "error: Cannot rename a non-local definition.",
+        )
+    }
 }

--- a/crates/ide/src/rename.rs
+++ b/crates/ide/src/rename.rs
@@ -2529,6 +2529,7 @@ fn main() {
 ",
         )
     }
+<<<<<<< HEAD
 
     #[test]
     fn extern_crate() {
@@ -2634,4 +2635,21 @@ use qux as frob;
         // ",
         //         );
     }
+||||||| parent of 948d9f274 (Disallow renaming of non-local structs)
+=======
+
+    #[test]
+    fn disallow_renaming_for_non_local_struct() {
+        check(
+            "Baz",
+            r#"
+//- /lib.rs crate:lib new_source_root:library
+pub struct S$0;
+//- /main.rs crate:main deps:lib new_source_root:local
+use lib::S;
+"#,
+            "error: Cannot rename a non-local struct.",
+        );
+    }
+>>>>>>> 948d9f274 (Disallow renaming of non-local structs)
 }

